guix: Use LTO to build releases

fanquake opened the pull request:
Changes required to use LTO in Guix (release) builds.

fanquake committed:
guix: build depends with LTO=1 and -flto

fanquake committed:
scripts: update exports for LTO

```bash
bitcoind: export of symbol in6addr_any not allowed!
bitcoind: failed EXPORTED_SYMBOLS
test/test_bitcoin: export of symbol in6addr_any not allowed!
test/test_bitcoin: failed EXPORTED_SYMBOLS
```

Also new exports for bitcoin-qt.

laanwj commented:
Concept ACK. This would be great.

hebasto commented:
> ... build all binaries except for `bitcoin-qt`

See #25542.

hebasto commented:
Tested 076aa5ee3e9d86cc5f0b4fdfe8926a7f6a7078a2 on Ubuntu 22.04:

```
$ make -C depends HOST=riscv64-linux-gnu LTO=1 NO_QT=1
$ ./autogen.sh
$ ./configure CONFIG_SITE=depends/riscv64-linux-gnu/share/config.site
$ make clean
$ make
...
  CXXLD    libbitcoinconsensus.la
/usr/lib/gcc-cross/riscv64-linux-gnu/11/../../../../riscv64-linux-gnu/bin/ld: secp256k1/.libs/libsecp256k1.a(libsecp256k1_la-secp256k1.o): Relocations in generic ELF (EM: 62)
/usr/lib/gcc-cross/riscv64-linux-gnu/11/../../../../riscv64-linux-gnu/bin/ld: secp256k1/.libs/libsecp256k1.a(libsecp256k1_la-secp256k1.o): Relocations in generic ELF (EM: 62)
/usr/lib/gcc-cross/riscv64-linux-gnu/11/../../../../riscv64-linux-gnu/bin/ld: secp256k1/.libs/libsecp256k1.a(libsecp256k1_la-secp256k1.o): Relocations in generic ELF (EM: 62)
/usr/lib/gcc-cross/riscv64-linux-gnu/11/../../../../riscv64-linux-gnu/bin/ld: secp256k1/.libs/libsecp256k1.a(libsecp256k1_la-secp256k1.o): Relocations in generic ELF (EM: 62)
/usr/lib/gcc-cross/riscv64-linux-gnu/11/../../../../riscv64-linux-gnu/bin/ld: secp256k1/.libs/libsecp256k1.a(libsecp256k1_la-secp256k1.o): Relocations in generic ELF (EM: 62)
/usr/lib/gcc-cross/riscv64-linux-gnu/11/../../../../riscv64-linux-gnu/bin/ld: secp256k1/.libs/libsecp256k1.a(libsecp256k1_la-secp256k1.o): Relocations in generic ELF (EM: 62)
/usr/lib/gcc-cross/riscv64-linux-gnu/11/../../../../riscv64-linux-gnu/bin/ld: secp256k1/.libs/libsecp256k1.a(libsecp256k1_la-secp256k1.o): Relocations in generic ELF (EM: 62)
/usr/lib/gcc-cross/riscv64-linux-gnu/11/../../../../riscv64-linux-gnu/bin/ld: secp256k1/.libs/libsecp256k1.a(libsecp256k1_la-secp256k1.o): Relocations in generic ELF (EM: 62)
/usr/lib/gcc-cross/riscv64-linux-gnu/11/../../../../riscv64-linux-gnu/bin/ld: secp256k1/.libs/libsecp256k1.a: error adding symbols: file in wrong format
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:6729: libbitcoinconsensus.la] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory '/home/hebasto/git/bitcoin/src'
make[1]: *** [Makefile:18615: all-recursive] Error 1
make[1]: Leaving directory '/home/hebasto/git/bitcoin/src'
make: *** [Makefile:818: all-recursive] Error 1
```

hebasto commented on depends/packages/expat.mk:
```diff
@@ -9,6 +9,7 @@ define $(package)_set_vars
   $(package)_config_opts += --disable-dependency-tracking --enable-option-checking
   $(package)_config_opts += --without-xmlwf
   $(package)_config_opts_linux=--with-pic
+  $(package)_cflags += -fno-lto
```
I'm pretty sure the https://github.com/libexpat/libexpat/commit/cf3c2aa0842187495ebe5f1bdb5f4f66c6ec3eff commit causes this problem.

fanquake commented:
Rebased for #25542. All binaries for our Linux HOSTS now build.

fanquake commented:
I've rebased on master for #25697 and added a couple commits that are not ideal, but make Windows Guix builds "work" for bitcoin-qt.exe. See the 8ddce8af136bf0c9908c7d8f690433de7d0df418 commit for details. It would be handy if someone with a Windows machine, maybe @hebasto or @jarolrod, could test the bitcoin-qt.exe binary. Will update Guix hashes in the PR description shortly.

hebasto commented:
> ... added a couple commits that are not ideal, but make Windows Guix builds "work" for bitcoin-qt.exe.

I guess it is related to this [bug](https://gcc.gnu.org/bugzilla//show_bug.cgi?id=94156) in `x86_64-w64-mingw32`.

hebasto commented:
> ... added a couple commits that are not ideal, but make Windows Guix builds "work" for bitcoin-qt.exe.

FWIW, there are some alternatives:
- disabling of accessibility, see my [`220727-qt-win-lto`](https://github.com/hebasto/bitcoin/commits/220727-qt-win-lto) branch (definitely a suboptimal option)
- using a non-gcc toolchain, e.g., LLVM-based?

fanquake commented:
Rebased on master / updated #25633. This now produces binaries for all HOSTS.

hebasto commented on depends/packages/qt.mk:
```diff
@@ -180,6 +180,9 @@ $(package)_config_opts_mingw32 += "QMAKE_LFLAGS = '$($(package)_ldflags)'"
 $(package)_config_opts_mingw32 += "QMAKE_LIB = '$($(package)_ar) rc'"
 $(package)_config_opts_mingw32 += -device-option CROSS_COMPILE="$(host)-"
 $(package)_config_opts_mingw32 += -pch
+ifneq ($(LTO),)
+$(package)_config_opts_mingw32 += -ltcg
+endif
```
FWIW, adding the `-ltcg` option saves 512 bytes only of `bitcoin-qt.exe` size being built in Guix.

Davidcalculus replied:
bc1qdrc8fup30ctv7ppjyr2txcg6wgcrejj2nk53we

jarolrod commented:
GUIX hashes

x86:
```

```

arm64:
```
$ env HOSTS='arm-linux-gnueabihf arm64-apple-darwin powerpc64-linux-gnu powerpc64le-linux-gnu riscv64-linux-gnu x86_64-apple-darwin x86_64-linux-gnu x86_64-w64-mingw32' ./contrib/guix/guix-build
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum

270ae9968f5189aa605b1c86a31806a72dd0b778979a27f8e55485b31a2ed0da  guix-build-39fc0b828ab2/output/arm-linux-gnueabihf/SHA256SUMS.part
4548af27e186766fa93bc6f731bb3462132f2316f92f9ae1feb70baedc377f92  guix-build-39fc0b828ab2/output/arm-linux-gnueabihf/bitcoin-39fc0b828ab2-arm-linux-gnueabihf-debug.tar.gz
e93f6284ac59263a3c38c8b823c85bb78208ec216c7411ad709919d18121a403  guix-build-39fc0b828ab2/output/arm-linux-gnueabihf/bitcoin-39fc0b828ab2-arm-linux-gnueabihf.tar.gz
ffaaabb6f35c908b4bd964d6ff081a653b9384a3150d33472bacb0374f5fe8d6  guix-build-39fc0b828ab2/output/arm64-apple-darwin/SHA256SUMS.part
93b9167ccf5aa79119b545e66073085474b5a9c2e7cb33c66525ae84dd19384c  guix-build-39fc0b828ab2/output/arm64-apple-darwin/bitcoin-39fc0b828ab2-arm64-apple-darwin-unsigned.dmg
cd3298fc409df17cc54ce958c3b5f9355e9e2945927ca4def47db4dbf13ca05a  guix-build-39fc0b828ab2/output/arm64-apple-darwin/bitcoin-39fc0b828ab2-arm64-apple-darwin-unsigned.tar.gz
f31493c0c2b661bf0e40eebd803b00dd72aa39bfec01d7aafd2c16c142e341da  guix-build-39fc0b828ab2/output/arm64-apple-darwin/bitcoin-39fc0b828ab2-arm64-apple-darwin.tar.gz
a31b401e0dc3e86c9d92836331f8f02752f0eb11d1570c9f6e799222f133548a  guix-build-39fc0b828ab2/output/dist-archive/bitcoin-39fc0b828ab2.tar.gz
947f01f6f4193cb9cc19b1349889f5d5599be3a5fd333023e8a00f746ecc87a9  guix-build-39fc0b828ab2/output/powerpc64-linux-gnu/SHA256SUMS.part
91ade76f7d594a6df9eb0528c8218ac0d0ff4baf76aab33c03dac3aacd0e2d98  guix-build-39fc0b828ab2/output/powerpc64-linux-gnu/bitcoin-39fc0b828ab2-powerpc64-linux-gnu-debug.tar.gz
f509a3858c646cd61dee16ad7fd50d841b95b22bf4888ccd09a4049d1bb166eb  guix-build-39fc0b828ab2/output/powerpc64-linux-gnu/bitcoin-39fc0b828ab2-powerpc64-linux-gnu.tar.gz
6236c96090cebc6ca1b7dae6681dc2044284169ffb9facb2f2c338a38a278e42  guix-build-39fc0b828ab2/output/powerpc64le-linux-gnu/SHA256SUMS.part
18333be581051c102d38480b750b879101983a2a54e528090849e9b1da521b60  guix-build-39fc0b828ab2/output/powerpc64le-linux-gnu/bitcoin-39fc0b828ab2-powerpc64le-linux-gnu-debug.tar.gz
1baeaf2b2aa4ae3197335302b8a2d1c62a45f74e99f5ca274e6980cceb956cb1  guix-build-39fc0b828ab2/output/powerpc64le-linux-gnu/bitcoin-39fc0b828ab2-powerpc64le-linux-gnu.tar.gz
f6208df8318fdee99fef5928e6fe46d5eb48e22a000be557e6ae57c77f7a94a7  guix-build-39fc0b828ab2/output/riscv64-linux-gnu/SHA256SUMS.part
2db3b2d30f0bff4e1b2ef06fe18532c9f5f64a9635ff1318427c0d8f0b365e1b  guix-build-39fc0b828ab2/output/riscv64-linux-gnu/bitcoin-39fc0b828ab2-riscv64-linux-gnu-debug.tar.gz
00acbb3f314443ce7e2660645b3afad29b670fb3a8780cdb15f8dc711c7c210e  guix-build-39fc0b828ab2/output/riscv64-linux-gnu/bitcoin-39fc0b828ab2-riscv64-linux-gnu.tar.gz
80e613f9fd970357f7794437bbd8f6b1884850ccebb1e6d4b4d56b646254c9bb  guix-build-39fc0b828ab2/output/x86_64-apple-darwin/SHA256SUMS.part
09da205c47c92ec730e88d7417acb49d24ce0fea952397460e23f2d07ee97e79  guix-build-39fc0b828ab2/output/x86_64-apple-darwin/bitcoin-39fc0b828ab2-x86_64-apple-darwin-unsigned.dmg
b7e433e3b304188b15660df5a1e47b69246cd53105aef492c68f11243efd54a3  guix-build-39fc0b828ab2/output/x86_64-apple-darwin/bitcoin-39fc0b828ab2-x86_64-apple-darwin-unsigned.tar.gz
d5360f1f5271651fecdaa1ba9b623336d8f76746c90ab07aee3a7cc83a3fb094  guix-build-39fc0b828ab2/output/x86_64-apple-darwin/bitcoin-39fc0b828ab2-x86_64-apple-darwin.tar.gz
98fb66b93499538643d7241c20327820e1ad51c406f947641ee686d1ab5b40f4  guix-build-39fc0b828ab2/output/x86_64-linux-gnu/SHA256SUMS.part
f32442a63506fd94eb72b883f3f4fc9ae6bc034be0673feec6234cf793e2890a  guix-build-39fc0b828ab2/output/x86_64-linux-gnu/bitcoin-39fc0b828ab2-x86_64-linux-gnu-debug.tar.gz
1b9c17ce72253416eff46c0fb9e5e1ca636ab019f39c3d6fa5a4d4d7bc9f81a6  guix-build-39fc0b828ab2/output/x86_64-linux-gnu/bitcoin-39fc0b828ab2-x86_64-linux-gnu.tar.gz
1d634370c83afc1027c28e46d5888edff050daa2c4b192c4c62e38266f2baf84  guix-build-39fc0b828ab2/output/x86_64-w64-mingw32/SHA256SUMS.part
b6ee25f1f9a869f7250e2708ae064c6006561c76e384f2c05b25ebab2f3861e5  guix-build-39fc0b828ab2/output/x86_64-w64-mingw32/bitcoin-39fc0b828ab2-win64-debug.zip
b1043edd2f1b2a4120a44422b6294b12939876a3a1c86f5e5474a4930fa1d344  guix-build-39fc0b828ab2/output/x86_64-w64-mingw32/bitcoin-39fc0b828ab2-win64-setup-unsigned.exe
60bcd0cdec77a30445509817a8b1befd9a9d1e4727faf27f5a729919588155c0  guix-build-39fc0b828ab2/output/x86_64-w64-mingw32/bitcoin-39fc0b828ab2-win64-unsigned.tar.gz
4eac6ccee5060f2392591c5a266490a027701c7ae057b6815025ccc2c0597c81  guix-build-39fc0b828ab2/output/x86_64-w64-mingw32/bitcoin-39fc0b828ab2-win64.zip
```

fanquake commented:
Rebased, and put on top of #25861. Please review that PR first.

hebasto reviewed (commented):
Approach ACK 80dfacbf81036a6fdbb033a8d7a169658a606751.

fanquake commented:
The non-determinism when building on x86_64 and aarch64 has now been fixed. This required backporting 1 commit to our gcc 10: https://github.com/bitcoin/bitcoin/pull/25391/commits/ca17654f507f2f0ff51f7a2802113c45331a8447 and one change in depends.

Returning this to draft, as I'll split out some of the prerequisistes. However general review, benchmarks, commentary would still be appreciated. I'll add Guix build hashes to the PR description shortly.

TheCharlatan commented:
Concept ACK

The size benefits for the windows builds are similar to the linux builds described in #25542. These here are cross-compiled with depends:
bitcoin-qt.exe size: 35.3MB
bitcoin-qt.exe LTO size: 30.9MB

The configure execution time is annoyingly slow, on my machine a normal configure run for cross-compiling w64 clocks in at ~16s, while with LTO enabled it clocks 5min16s. The reason for this is the Qt plugin check. It links all the other detected Qt libraries with the static plugin library. I tried speeding this check up by optimizing the selection of linked libraries, but ended up with a not significantly faster solution. Would it make sense to drop the LTO flag for this test instead?

When building `bitcoind.exe`, I see the following warning:
```
/usr/share/mingw-w64/include/psdk_inc/_fd_types.h:21: note: array types have different bounds
   21 | typedef struct fd_set
      | 
  AR       libtest_fuzz.a
/usr/share/mingw-w64/include/psdk_inc/_fd_types.h:21: note: array types have different bounds
   21 | typedef struct fd_set
      | 
  CXXLD    test/fuzz/fuzz.exe
/usr/share/mingw-w64/include/psdk_inc/_fd_types.h:21: note: array types have different bounds
   21 | typedef struct fd_set
      | 
/usr/share/mingw-w64/include/psdk_inc/_fd_types.h:21: note: array types have different bounds
   21 | typedef struct fd_set
      | 
```
There are also a bunch of :
```
../depends/work/build/x86_64-w64-mingw32/miniupnpc/2.2.2-2b1a5331a74/miniupnpc.c: In function 'GetUPNPUrls':
../depends/work/build/x86_64-w64-mingw32/miniupnpc/2.2.2-2b1a5331a74/miniupnpc.c:496:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
```

TheCharlatan commented:
Guix build

```
1b8c738d978810a1ef6af91e9bee4a60368de1f1669420cd29c40e2f5d322a82  output/aarch64-linux-gnu/bitcoin-5f7a0db6480b-aarch64-linux-gnu-debug.tar.gz
dbcbbf4c037ee664d7e427725adfdff59e8c788d9da949baa653e363e07be82d  output/aarch64-linux-gnu/bitcoin-5f7a0db6480b-aarch64-linux-gnu.tar.gz
0d49a7934d2ffd9d497b0e8f87e4ca90b1c516fba273fff963eb8baf0dddb73d  output/aarch64-linux-gnu/SHA256SUMS.part
669451b707d585f056dc4cc498191890f7b581b03e344852c288026f23c8750b  output/arm64-apple-darwin/bitcoin-5f7a0db6480b-arm64-apple-darwin.tar.gz
5e394962f842771dff13821140864ebf881bef4242d12bef06363d9ee5e41d89  output/arm64-apple-darwin/bitcoin-5f7a0db6480b-arm64-apple-darwin-unsigned.dmg
bb45f07390fb657bcfd9727782eb75056e49cc0796d0d24f51ef96f30ddfe269  output/arm64-apple-darwin/bitcoin-5f7a0db6480b-arm64-apple-darwin-unsigned.tar.gz
8ba2ecb625e2cc178a3a6f0ccde390a71ee8dbc24df7d2f4affc386ca6b1b2cf  output/arm64-apple-darwin/SHA256SUMS.part
7a1323bc756260e3f6212abaf3c55d0aac74ec2fefbb48e1ba0accb956b4e60c  output/arm-linux-gnueabihf/bitcoin-5f7a0db6480b-arm-linux-gnueabihf-debug.tar.gz
d6905023f4c2071029e04ad331c76c2219c85029be152286344fd6364e1dff48  output/arm-linux-gnueabihf/bitcoin-5f7a0db6480b-arm-linux-gnueabihf.tar.gz
9b0f34c730aecd20aa03ad8d091a705b203e10d2b4f428b6fa9be22e8b783280  output/arm-linux-gnueabihf/SHA256SUMS.part
fe52699b528416963d52f9243a80113524ea536d54b8ca12224d8294e9cb98ed  output/dist-archive/bitcoin-5f7a0db6480b.tar.gz
ceb0da9200dc81e0e43c18ffe8987a4ae0ef465eb1968c1e69ccf9ff9bec015b  output/powerpc64le-linux-gnu/bitcoin-5f7a0db6480b-powerpc64le-linux-gnu-debug.tar.gz
5dffc49e0dd101ee691ceccd63a8efdec2e2189a572ed5254125774f7dfcc0a4  output/powerpc64le-linux-gnu/bitcoin-5f7a0db6480b-powerpc64le-linux-gnu.tar.gz
9767c0ea1823ef5475ea1eaf3127a594b7d67c6bc8b9cb4aab4e38a8a3f20f1e  output/powerpc64le-linux-gnu/SHA256SUMS.part
00f92b29ba66ba77038a5cf9c45029b5883dce705262674defb0668fabd1dfa0  output/powerpc64-linux-gnu/bitcoin-5f7a0db6480b-powerpc64-linux-gnu-debug.tar.gz
8dad945cc069c67549e6cbbf5a94812f21a786208c5c7a47d8f84756cc483bf2  output/powerpc64-linux-gnu/bitcoin-5f7a0db6480b-powerpc64-linux-gnu.tar.gz
41e1f05fe67a827df4d9c4f829cbad2408d93ec6f8e13c7c32a41d226496ae21  output/powerpc64-linux-gnu/SHA256SUMS.part
f36781a02f81a3822178857d644fb69d30d7e0c614dfda4d5d945e5b9aa69279  output/riscv64-linux-gnu/bitcoin-5f7a0db6480b-riscv64-linux-gnu-debug.tar.gz
c1a3c493f5c87ace3d5215c24a7a30dab1e517ad5acc1bf85c3651a7584074f9  output/riscv64-linux-gnu/bitcoin-5f7a0db6480b-riscv64-linux-gnu.tar.gz
4dbe3a6d6587b13ffedba2c88012a9febaef7039762788f8106f9f3776b8e071  output/riscv64-linux-gnu/SHA256SUMS.part
9a7a65e4b733df36b19b12e808ab89b3f78f0535b378b631f6dd7b07d92eea21  output/x86_64-apple-darwin/bitcoin-5f7a0db6480b-x86_64-apple-darwin.tar.gz
04b553a9886ba214be47d4063f8503abd281e1a15bdafed9b7f959d6d377c48f  output/x86_64-apple-darwin/bitcoin-5f7a0db6480b-x86_64-apple-darwin-unsigned.dmg
c231d76279e95459eaf7913fca01ef3091645ebc892d30ffdbeb77976905dc2b  output/x86_64-apple-darwin/bitcoin-5f7a0db6480b-x86_64-apple-darwin-unsigned.tar.gz
3bffcc0916498eb65656c41727b0d5247830513daa705678e4a47dc8bc1cbe6a  output/x86_64-apple-darwin/SHA256SUMS.part
d16134283a0e82274a54e747eaaf4173e9fab06e9cf3a255a4e0ce2a33575931  output/x86_64-linux-gnu/bitcoin-5f7a0db6480b-x86_64-linux-gnu-debug.tar.gz
ab9a2f28eccf8ddf355e0ffedc844b06c514b0bee2bf584dd4fb065951683156  output/x86_64-linux-gnu/bitcoin-5f7a0db6480b-x86_64-linux-gnu.tar.gz
f49b261c6461f647c9639ff9522f85420f7a237af2e580adf8814321be8e4d98  output/x86_64-linux-gnu/SHA256SUMS.part
c37bb94d1214e2dd12f6edbcf4d17810b6d934ca9f4248c6bb48c76e1837c6c8  output/x86_64-w64-mingw32/bitcoin-5f7a0db6480b-win64-debug.zip
bbb7c9e1ec260f8f1b562f76a90ea68ff6cea1b2c06e32c8d6e3ccea3aa6cde7  output/x86_64-w64-mingw32/bitcoin-5f7a0db6480b-win64-setup-unsigned.exe
c534e9629515740c66102326ec9727a33ec8a4e626a86d755ce76d28318b00b8  output/x86_64-w64-mingw32/bitcoin-5f7a0db6480b-win64-unsigned.tar.gz
21abc5257ddcf02e2e99269edc91678d57862e325939f24031b5e134190d7a18  output/x86_64-w64-mingw32/bitcoin-5f7a0db6480b-win64.zip
5dd91aef430c528db8779d3e6fa112eb8d00a72739b8cd61a1a08254c275bd5e  output/x86_64-w64-mingw32/SHA256SUMS.part
```

fanquake commented:
Rebased past #27153.

fanquake commented:
Rebased this on top of #27897. Dropped the commit to backport a patch to GCC 10.

fanquake commented:
Rebased on #27897. Simplifed / squashed down the changes here. Updated the PR description. The current (non-conceptual) blocker is that Qt for Windows doesn't really work under LTO... See 8031e400af66bf5267201069a7c062cb6a9d02ea.

theuni commented:
I have no idea if this will help, but this should keep the dtor in a single compilation unit.

```patch
diff -ruN qtbase-everywhere-src-5.15.2/src/widgets/accessible/simplewidgets.cpp qtbase-everywhere-src-5.15.2.new/src/widgets/accessible/simplewidgets.cpp
--- qtbase-everywhere-src-5.15.2/src/widgets/accessible/simplewidgets.cpp	2020-10-27 08:02:11.000000000 +0000
+++ qtbase-everywhere-src-5.15.2.new/src/widgets/accessible/simplewidgets.cpp	2023-09-20 13:58:14.878755026 +0000
@@ -678,6 +678,8 @@
   \ingroup accessibility
 */

+QAccessibleLineEdit::~QAccessibleLineEdit();
+
 /*!
   Constructs a QAccessibleLineEdit object for \a w.
   \a name is propagated to the QAccessibleWidget constructor.
diff -ruN qtbase-everywhere-src-5.15.2/src/widgets/accessible/simplewidgets_p.h qtbase-everywhere-src-5.15.2.new/src/widgets/accessible/simplewidgets_p.h
--- qtbase-everywhere-src-5.15.2/src/widgets/accessible/simplewidgets_p.h	2020-10-27 08:02:11.000000000 +0000
+++ qtbase-everywhere-src-5.15.2.new/src/widgets/accessible/simplewidgets_p.h	2023-09-20 13:52:10.467865415 +0000
@@ -154,7 +154,7 @@
 {
 public:
     explicit QAccessibleLineEdit(QWidget *o, const QString &name = QString());
-
+    ~QAccessibleLineEdit();
     QString text(QAccessible::Text t) const override;
     void setText(QAccessible::Text t, const QString &text) override;
     QAccessible::State state() const override;
```

fanquake commented:
> I have no idea if this will help,

Looks like it's good enough for working Windows builds. Turned the change into a depends patch, and dropped `--allow-multiple-definition`. I'll PR this separately, as we can fix the `LTO=1` case for Windows, before turning this on for releases.

theuni commented:
Any idea where `in6addr_any` is coming from? Seems fine, just curious.

fanquake commented:
> Any idea where in6addr_any is coming from? Seems fine, just curious.

I can't remember exactly, but it looks like this is actually no-longer required. Did a new round of builds with this branch, and it dropped from the script.

hebasto commented:
Related:
- https://github.com/bitcoin/bitcoin/pull/29185

fanquake commented:
> Would be nice to extend the motivation (pull request description) a bit about the benefits and risks.

Thanks. Drafted for the moment, but going to come back here soon with much more information.

fanquake commented:
Going to come back to this after static builds.
